Formalist/FormNumberFormatter.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  FormNumberFormatter.swift
+//  Formalist
+//
+//  Created by Viktor Radchenko on 7/21/17.
+//  Copyright © 2017 Seed Platform, Inc. All rights reserved.
+//
+
+public enum FormNumberType {
+    case EIN
+    case SSN
+    case creditCard
+    case custom(pattern: String, replaceCharacter: Character)
+}
+
+public struct FormNumberFormatter: Formattable {
+
+    let type: FormNumberType
+
+    public init(
+        type: FormNumberType
+    ) {
+        self.type = type
+    }
+
+    public func from(input: String, previousInput: String = "") -> String {
+        let newValue = input.digits
+        var result = ""
+        let pattern = type.pattern
+        let replaceCharacter = type.replaceCharacter
+        var newValueIndex = newValue.startIndex
+        var patternIndex = pattern.startIndex
+        let isRemoving = newValue.characters.count < previousInput.characters.count
+
+        while patternIndex < pattern.endIndex && newValueIndex < newValue.endIndex  {
+            let value = pattern[patternIndex]
+
+            if value == replaceCharacter {
+                result.append(newValue[newValueIndex])
+                newValueIndex = newValue.index(newValueIndex, offsetBy: 1)
+            } else {
+                result.append(value)
+            }
+            patternIndex = pattern.index(patternIndex, offsetBy: 1)
+
+            if newValueIndex == newValue.endIndex && patternIndex < pattern.endIndex && pattern[patternIndex] != replaceCharacter {
+                if isRemoving {
```

Comment:
I think this still causes some weirdness. Try typing `333` and you end up with `333-`, then hit any non-number and you get `33`. It is probably to just remove this code and have `333` display as `333` and have the dash shown when you enter the next number. This way you don't need to track `isRemoving` to determine whether to add another or remove a character.